### PR TITLE
Output column number info when panicking

### PR DESCRIPTION
**Pull request by est31**
Outputs the column number when panicking. Useful when you e.g. have code like `foo[i] = bar[k] + bar[l]` and you get a panic with index out of bounds, or when you have an expression like `a = b + c + d + e` and the addition overflows. Now you know which operation to blame!

The format is `file:line:column`, just like for compiler errors. Example output with the patch:

```
thread 'main' panicked at 'index out of bounds: the len is 5 but the index is 8', src/main.rs:3:8
```

As some of the API between the compiler and the library landscape gets broken, this is a bit hackier than I'd originally wanted it to be.

* `panic` and `panic_bounds_check` lang items got an additional column param, on stage0 I still have to use the previous version. After a SNAP this should be resolved.
* For `#[derive(RustcDeserialze)]`, stage0 requires a fixed signature for `std::rt::begin_panic`, so we can't change it right away. What we need to do instead is to keep the signature, and add a `begin_panic_new` function that we use in later stages instead. After a SNAP we can change the `begin_panic` function and rely on it instead of `begin_panic_new`, and one SNAP later we can remove `begin_panic_new`.
* Fortunately I didn't have to break anything about the panic hook API, I could easily extend it.

Note that debuginfo remains unchanged, so RUST_BACKTRACE output won't contain any column info. See issue #42921 for discussion on including the column in debuginfo.

### CODE REVIEW

**Comment by eddyb** [comment]
cc @rust-lang/libs About the addition of `panicking::Location::column`. r=me otherwise.

**Comment by est31** [comment]
I've renamed the commit (without any code changes) and added a second one with a small refactor, a comment for the stage0 updater, and (hopefully) a fix for the test failure.

**Comment by sfackler** [comment]
:+1:

**Comment by est31** [comment]
ready. r? @eddyb

**Comment by est31** [comment]
cc @alexcrichton

File: src/libcore/panicking.rs [hunk 1]
```diff
@@ -39,34 +39,55 @@
...
 
 #[cold] #[inline(never)]
-pub fn panic_fmt(fmt: fmt::Arguments, file_line: &(&'static str, u32)) -> ! {
+pub fn panic_fmt(fmt: fmt::Arguments, file_line_col: &(&'static str, u32, u32)) -> ! {
     #[allow(improper_ctypes)]
     extern {
         #[lang = "panic_fmt"]
         #[unwind]
-        fn panic_impl(fmt: fmt::Arguments, file: &'static str, line: u32) -> !;
+        fn panic_impl(fmt: fmt::Arguments, file: &'static str, line: u32, col :u32) -> !;
```
Comment (alexcrichton) [inline src/libcore/panicking.rs]: s/col :u32/col: u32/

**Review by alexcrichton — COMMENTED** [review]
Ok looks pretty good to me, thanks @est31! Mind doing some size analysis to ensure this doesn't increase the size of anything too much?

**Comment by est31** [comment]
> Mind doing some size analysis to ensure this doesn't increase the size of anything too much?

I've compiled rustc with the patch and with the commit this PR bases on, and these are the results:

```

Without patch:
120136	build/x86_64-unknown-linux-gnu/stage1-rustc
93752	build/x86_64-unknown-linux-gnu/stage1-std

With patch:
120148	build/x86_64-unknown-linux-gnu/stage1-rustc
93772	build/x86_64-unknown-linux-gnu/stage1-std

```

Numbers come from du output, and are therefore in kb.

So almost no noticeable change, its in the sub-0.1% range for both std and rustc.

**Comment by est31** [comment]
The failure looks legit, I'll investigate.

**Comment by est31** [comment]
The error is inside cargo's testsuite:

<details><summary>(expand)</summary>

```

[01:11:06] thread 'cargo_bench_failing_test' panicked at '
[01:11:06] Expected: execs
[01:11:06]     but: expected to find:
[01:11:06] [..]src[/]foo.rs:14
[01:11:06] 
[01:11:06] did not find in output:
[01:11:06]    Compiling foo v0.5.0 (file:///checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/cit/t11/foo)
[01:11:06]     Finished release [optimized] target(s) in 0.67 secs
[01:11:06]      Running target/release/deps/foo-b865dce48c3e4e89
[01:11:06] thread 'main' panicked at 'assertion failed: `(left == right)`
[01:11:06]   left: `"hello"`,
[01:11:06]  right: `"nope"`', src/foo.rs:14:16
[01:11:06] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:11:06] error: bench failed
[01:11:06] ', /cargo/registry/src/github.com-1ecc6299db9ec823/hamcrest-0.1.1/src/core.rs:31:12
```

</details>

**Comment by est31** [comment]
Fixed & rebased due to merge conflicts. re-r? @eddyb

**Comment by Mark-Simulacrum** [comment]
I don't see an update to Cargo.toml in this PR, which I believe is needed -- we store the git sha for Cargo in there as well, IIRC.

```
[01:12:30] ---- cargo_test_failing_test_in_lib stdout ----
[01:12:30] 	running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/cargo test`
[01:12:30] thread 'cargo_test_failing_test_in_lib' panicked at '
[01:12:30] Expected: execs
[01:12:30]     but: expected to find:
[01:12:30] test test_hello ... FAILED
[01:12:30] 
[01:12:30] failures:
[01:12:30] 
[01:12:30] ---- test_hello stdout ----
[01:12:30] <tab>thread 'test_hello' panicked at 'assertion failed: false', src[/]lib.rs:4
[01:12:30] 
[01:12:30] 
[01:12:30] did not find in output:
[01:12:30] 
[01:12:30] running 1 test
[01:12:30] test test_hello ... FAILED
[01:12:30] 
[01:12:30] failures:
[01:12:30] 
[01:12:30] ---- test_hello stdout ----
[01:12:30] <tab>thread 'test_hello' panicked at 'assertion failed: false', src/lib.rs:4:16
[01:12:30] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:12:30] 
[01:12:30] 
[01:12:30] failures:
[01:12:30]     test_hello
[01:12:30] 
[01:12:30] test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
[01:12:30] 
[01:12:30] ', /cargo/registry/src/github.com-1ecc6299db9ec823/hamcrest-0.1.1/src/core.rs:31:12
[01:12:30] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:12:30] 
[01:12:30] ---- cargo_test_failing_test_in_test stdout ----
[01:12:30] 	running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/cargo build`
[01:12:30] running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/cit/t11/foo/target/debug/foo`
[01:12:30] running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/cargo test`
[01:12:30] thread 'cargo_test_failing_test_in_test' panicked at '
[01:12:30] Expected: execs
[01:12:30]     but: expected to find:
[01:12:30] running 1 test
[01:12:30] test test_hello ... FAILED
[01:12:30] 
[01:12:30] failures:
[01:12:30] 
[01:12:30] ---- test_hello stdout ----
[01:12:30] <tab>thread 'test_hello' panicked at 'assertion failed: false', tests[/]footest.rs:4
[01:12:30] 
[01:12:30] 
[01:12:30] did not find in output:
[01:12:30] 
[01:12:30] running 0 tests
[01:12:30] 
[01:12:30] test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
[01:12:30] 
[01:12:30] 
[01:12:30] running 1 test
[01:12:30] test test_hello ... FAILED
[01:12:30] 
[01:12:30] failures:
[01:12:30] 
[01:12:30] ---- test_hello stdout ----
[01:12:30] <tab>thread 'test_hello' panicked at 'assertion failed: false', tests/footest.rs:4:16
[01:12:30] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:12:30] 
[01:12:30] 
[01:12:30] failures:
[01:12:30]     test_hello
[01:12:30] 
[01:12:30] test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
[01:12:30] 
[01:12:30] ', /cargo/registry/src/github.com-1ecc6299db9ec823/hamcrest-0.1.1/src/core.rs:31:12
[01:12:30] 
[01:12:30] 
[01:12:30] failures:
[01:12:30]     cargo_test_failing_test_in_lib
[01:12:30]     cargo_test_failing_test_in_test
```

**Comment by est31** [comment]
@Mark-Simulacrum are you sure that an update to Cargo.toml/Cargo.lock is needed? E.g. look at git commit 77931c2a04fbb41f3cd92435aa5294222bb5db7f, it has no update.

About the other cargo failures... I'll try to fix them as well. Right now my hardware is a bit limited so I can't really run the testsuite sadly...

**Comment by Mark-Simulacrum** [comment]
Cargo.lock was updated though: https://github.com/rust-lang/rust/commit/77931c2a04fbb41f3cd92435aa5294222bb5db7f#diff-5d9bf57935c52008319cccc88d4d737dR160.

**Pull request merged**